src/main/java/org/jboss/tm/listener/event/TransactionLifecycleEventType.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2017, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.tm.listener.event;
+
+import java.util.EnumSet;
+
+/**
+ * A mechanism for providing extensible enum types.
+ *
+ * @param <E> the Enum type of the implementing class - ie the implementation will have all of the methods of
+ *           Enum<E>
+ */
+public interface TransactionLifecycleEventType<E extends Enum<E>> {
```

Comment:
I'm now re-thinking this again. I would like understand better what's behind the decision of having defined `TransactionLifecycleEventType` as parent class of all event. I missed that previously. Don't you mind giving me a context, plase?

My doubt is if it happens that one listener needs to reuse an event type of another listener? I can see here that `vetoed` listener and `associated` listener use its own even type (`AssociationLifecycleEventType` vs. `TransactionLifecycleEventType`) and do not seem to need to reuse the listener from the other one. I mean if it's needed to join all events under the `TransactionLifecycleEventType`.

The current code uses `EnumSet` and adding another `Enum` with e.g. vetoing event is way how to enum types for events could be extended, or no? I mean why simply not just add another `Enum` type? (I know that you mention exactly this at https://issues.jboss.org/browse/JBTM-2863 but as I do not understand I'm asking)